Version 1.0.0

# Introduction

In this programming assignment we will illustrate a very severe data leakage, that can often be found in competitions, where the pairs of object should be scored, e.g. predict $1$ if two objects belong to the same class and $0$ otherwise. 

The data in this assignment is taken from a real competition, and the funniest thing is that *we will not use training set at all* and achieve almost 100% accuracy score! We will just exploit the leakage.

Now go through the notebook and complete the assignment.

In [1]:
import numpy as np
import pandas as pd 
import scipy.sparse

# Load the data

Let's load the test data. Note, that we don't have any training data here, just test data. Moreover, *we will not even use any features* of test objects. All we need to solve this task is the file with the indices for the pairs, that we need to compare.

Let's load the data with test indices.

In [2]:
test = pd.read_csv('test_pairs.csv')
test.head(10)

pairId  FirstId  SecondId
0       0     1427      8053
1       1    17044      7681
2       2    19237     20966
3       3     8005     20765
4       4    16837       599
5       5     3657     12504
6       6     2836      7582
7       7     6136      6111
8       8    23295      9817
9       9     6621      7672

For example, we can think that there is a test dataset of images, and each image is assigned a unique `Id` from $0$ to $N-1$ (N -- is the number of images). In the dataframe from above `FirstId` and `SecondId` point to these `Id`'s and define pairs, that we should compare: e.g. do both images in the pair belong to the same class or not. So, for example for the first row: if images with `Id=1427` and `Id=8053` belong to the same class, we should predict $1$, and $0$ otherwise. 

But in our case we don't really care about the images, and how exactly we compare the images (as long as comparator is binary).  

**We suggest you to try to solve the puzzle yourself first.** You need to submit a `.csv` file with columns `pairId` and `Prediction` to the grader. The number of submissions allowed is made pretty huge to let you explore the data without worries. The returned score should be very close to $1$.

**If you do not want to think much** -- scroll down and follow the instructions below.

In [7]:
test['SecondId'].value_counts()

12194    23
12223    22
12279    22
12162    22
12141    22
         ..
26121     1
26126     1
26118     1
26125     1
26122     1
Name: SecondId, Length: 26310, dtype: int64

# EDA and leakage intuition

As we already know, the key to discover data leakages is careful EDA. So let's start our work with some basic data exploration and build an intuition about the leakage.

First, check, how many different `id`s are there: concatenate `FirstId` and `SecondId` and print the number of unique elements. Also print minimum and maximum value for that vector.

In [9]:
# YOUR CODE GOES HERE
concated_id = test.apply(lambda row: str(row['FirstId'])+str(row['SecondId']), axis=1).astype(int)
concated_id

0           14278053
1          170447681
2         1923720966
3          800520765
4           16837599
             ...    
368545      29966659
368546     136263390
368547     254072932
368548     253055588
368549     176711874
Length: 368550, dtype: int64

In [11]:
concated_id.value_counts()

4213221       2
415510404     2
22953819      2
6107235       2
4981122       2
             ..
70884761      1
56808677      1
250711451     1
1897510300    1
8914945       1
Length: 368509, dtype: int64

and then print how many pairs we need to classify (it is basically the number of rows in the test set)

In [12]:
concated_id.nunique()

368509

In [45]:
test.max()

pairId      368549
FirstId      26324
SecondId     26324
dtype: int64

Now print, how many distinct pairs it would be possible to create out of all "images" in the dataset?   

In [13]:
# YOUR CODE GOES HERE
sorted_test = test.copy()
sorted_test.loc[:, ['FirstId', 'SecondId']] = np.sort(
    sorted_test.loc[:, ['FirstId', 'SecondId']].to_numpy(), axis=1)
unique_pairs, unique_counts = np.unique(
    sorted_test.loc[:, ['FirstId', 'SecondId']].to_numpy(), axis=0, return_counts=True)

In [14]:
unique_pairs[unique_counts > 1]

array([[  237, 11620],
       [  256,  2551],
       [  292,  2587],
       [  292,  4115],
       [  292,  6251],
       [  292,  8657],
       [  292,  9968],
       [  292, 12084],
       [  292, 12960],
       [  292, 15147],
       [  292, 16423],
       [  292, 18350],
       [  292, 19685],
       [  292, 21415],
       [  292, 23544],
       [  292, 25271],
       [  300,  2595],
       [  300,  4123],
       [  300,  6259],
       [  300,  8665],
       [  300,  9976],
       [  300, 12092],
       [  300, 12968],
       [  300, 15155],
       [  300, 16431],
       [  300, 18358],
       [  300, 19693],
       [  300, 21423],
       [  300, 23552],
       [  300, 25279],
       [  308,  1400],
       [  332, 10404],
       [  382, 11461],
       [  504,  2799],
       [  874,  6819],
       [ 1420,  8046],
       [ 1443, 12212],
       [ 1703,  7369],
       [ 1850,  9716],
       [ 1980,  8046],
       [ 2420,  6566],
       [ 2487, 11083],
       [ 2551,  4079],
       [ 25

So the number of pairs we are given to classify is very very small compared to the total number of pairs. 

To exploit the leak we need to **assume (or prove)**, that the total number of positive pairs is small, compared to the total number of pairs. For example: think about an image dataset with $1000$ classes, $N$ images per class. Then if the task was to tell whether a pair of images belongs to the same class or not, we would have $1000\frac{N(N-1)}{2}$ positive pairs, while total number of pairs was $\frac{1000N(1000N - 1)}{2}$.

Another example: in [Quora competitition](https://www.kaggle.com/c/quora-question-pairs) the task was to classify whether a pair of qustions are duplicates of each other or not. Of course, total number of question pairs is very huge, while number of duplicates (positive pairs) is much much smaller.

Finally, let's get a fraction of pairs of class `1`. We just need to submit a constant prediction "all ones" and check the returned accuracy. Create a dataframe with columns `pairId` and `Prediction`, fill it and export it to `.csv` file. Then submit to grader and examine grader's output. 

In [16]:
output = test[['pairId']]
output['Prediction'] = 1
output.set_index('pairId').to_csv('test_all_one.csv')

<ipython-input-16-ed9f9e6c82d9>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  output['Prediction'] = 1


So, we assumed the total number of pairs is much higher than the number of positive pairs, but it is not the case for the test set. It means that the test set is constructed not by sampling random pairs, but with a specific sampling algorithm. Pairs of class `1` are oversampled.

Now think, how we can exploit this fact? What is the leak here? If you get it now, you may try to get to the final answer yourself, othewise you can follow the instructions below.   

# Building a magic feature

In this section we will build a magic feature, that will solve the problem almost perfectly. The instructions will lead you to the correct solution, but please, try to explain the purpose of the steps we do to yourself -- it is very important.

## Incidence matrix

First, we need to build an [incidence matrix](https://en.wikipedia.org/wiki/Incidence_matrix). You can think of pairs `(FirstId, SecondId)` as of edges in an undirected graph. 

The incidence matrix is a matrix of size `(maxId + 1, maxId + 1)`, where each row (column) `i` corresponds `i-th` `Id`. In this matrix we put the value `1` to the position `[i, j]`, if and only if a pair `(i, j)` or `(j, i)` is present in  a given set of pais `(FirstId, SecondId)`. All the other elements in the incidence matrix are zeros.   

**Important!** The incidence matrices are typically very very sparse (small number of non-zero values). At the same time incidence matrices are usually huge in terms of total number of elements, and it is **impossible to store them in memory in dense format**. But due to their sparsity incidence matrices **can be easily represented as sparse matrices**. If you are not familiar with sparse matrices, please see [wiki](https://en.wikipedia.org/wiki/Sparse_matrix) and [scipy.sparse reference](https://docs.scipy.org/doc/scipy/reference/sparse.html). Please, use any of `scipy.sparse` constructors to build incidence matrix. 

For example, you can use this constructor: `scipy.sparse.coo_matrix((data, (i, j)))`. We highly recommend to learn to use different `scipy.sparse` constuctors, and matrices types, but if you feel you don't want to use them, you can always build this matrix with a simple `for` loop. You will need first to create a matrix using `scipy.sparse.coo_matrix((M, N), [dtype])` with an appropriate shape `(M, N)` and then iterate through `(FirstId, SecondId)` pairs and fill corresponding elements in matrix with ones. 

**Note**, that the matrix should be symmetric and consist only of zeros and ones. It is a way to check yourself.

In [25]:
same_pair_indices = unique_pairs[:, 0] == unique_pairs[:, 1]
row = np.hstack((unique_pairs[:, 0], unique_pairs[:, 1][~same_pair_indices]))
col = np.hstack((unique_pairs[:, 1], unique_pairs[:, 0][~same_pair_indices]))

inc_mat = scipy.sparse.coo_matrix((np.ones(len(row)), (row, col)))

# Sanity checks
assert inc_mat.max() == 1
assert inc_mat.sum() == 736872

In [30]:
row, col

(array([    0,     0,     0, ..., 26309, 26323, 26324]),
 array([  198,   552,  2299, ..., 26307, 26313, 26314]))

It is convenient to have matrix in `csr` format eventually.

In [32]:
inc_mat = inc_mat.tocsr()

In [33]:
inc_mat

<26325x26325 sparse matrix of type '<class 'numpy.float64'>'
	with 736872 stored elements in Compressed Sparse Row format>

## Now build the magic feature

Why did we build the incidence matrix? We can think of the rows in this matix as of representations for the objects. `i-th` row is a representation for an object with `Id = i`. Then, to measure similarity between two objects we can measure similarity between their representations. And we will see, that such representations are very good.

Now select the rows from the incidence matrix, that correspond to `test.FirstId`'s, and `test.SecondId`'s.

In [50]:
# Note, scipy goes crazy if a matrix is indexed with pandas' series. 
# So do not forget to convert `pd.series` to `np.array`
# These lines should normally run very quickly 

rows_FirstId   = test['FirstId'].to_numpy()
rows_SecondId  = test['SecondId'].to_numpy()

Our magic feature will be the *dot product* between representations of a pair of objects. Dot product can be regarded as similarity measure -- for our non-negative representations the dot product is close to 0 when the representations are different, and is huge, when representations are similar. 

Now compute dot product between corresponding rows in `rows_FirstId` and `rows_SecondId` matrices.

In [71]:
# Note, that in order to do pointwise multiplication in scipy.sparse you need to use function `multiply`
# regular `*` corresponds to matrix-matrix multiplication

f = inc_mat[rows_FirstId].multiply(inc_mat[rows_SecondId]).sum(axis=1)
f = np.squeeze(np.asarray(f))
# Sanity check
assert f.shape == (368550, )

That is it! **We've built our magic feature.** 

# From magic feature to binary predictions

But how do we convert this feature into binary predictions? We do not have a train set to learn a model, but we have a piece of information about test set: the baseline accuracy score that you got, when submitting constant. And we also have a very strong considerations about the data generative process, so probably we will be fine even without a training set. 

We may try to choose a thresold, and set the predictions to 1, if the feature value `f` is higer than the threshold, and 0 otherwise. What threshold would you choose? 

How do we find a right threshold? Let's first examine this feature: print frequencies (or counts) of each value in the feature `f`.

In [76]:
# For example use `np.unique` function, check for flags

print(np.unique(f))

[14. 15. 19. 20. 21. 28. 35.]


Do you see how this feature clusters the pairs? Maybe you can guess a good threshold by looking at the values? 

In fact, in other situations it can be not that obvious, but in general to pick a threshold you only need to remember the score of your baseline submission and use this information. Do you understand why and how?  

Choose a threshold below: 

In [81]:
pred = f >= 19

# Finally, let's create a submission

In [82]:
submission = test.loc[:,['pairId']]
submission['Prediction'] = pred.astype(int)

submission.to_csv('submission.csv', index=False)

Now submit it to the grader! It is not possible to submit directly from this notebook, as we need to submit a `csv` file, not a single number (limitation of Coursera platform). 

To download `submission.csv` file that you've just produced <a href='./submission.csv'>click here</a> (if the link opens in browser, right-click on it and shoose "Save link as"). Then go to [assignment page](https://www.coursera.org/learn/competitive-data-science/programming/KsASv/data-leakages/submission) and submit your `.csv` file in 'My submission' tab.


If you did everything right, the score should be very high.

**Finally:** try to explain to yourself, why the whole thing worked out. In fact, there is no magic in this feature, and the idea to use rows in the incidence matrix can be intuitively justified.

# Bonus

Interestingly, it is not the only leak in this dataset. There is another totally different way to get almost 100% accuracy. Try to find it!